File: platforms/android/zylix-android/src/main/java/com/zylix/ZylixRouter.kt
// ZylixRouter.kt - Android Router for Zylix v0.3.0
//
// Provides Jetpack Compose Navigation integration for Zylix routing system.
// Features:
// - NavHost/NavController integration
// - Deep link handling (App Links)
// - Route parameters and query strings
// - Navigation guards
// - Type-safe routing

package com.zylix

import android.net.Uri
import androidx.compose.runtime.*
import androidx.navigation.NavHostController
import androidx.navigation.compose.rememberNavController
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

// ============================================================================
// Route Definition
// ============================================================================

/**
 * A route parameter extracted from URL
 */
data class RouteParam(
    val name: String,
    val value: String
)

/**
 * Parsed URL components
 */
data class ParsedURL(
    val path: String,
    val params: List<RouteParam> = emptyList(),
    val query: Map<String, String> = emptyMap(),
    val fragment: String? = null
) {
    fun getParam(name: String): String? = params.find { it.name == name }?.value
    fun getQuery(key: String): String? = query[key]
}

/**
 * Route guard result
 */
sealed class GuardResult {
    object Allow : GuardResult()
    data class Deny(val message: String? = null) : GuardResult()
    data class Redirect(val to: String) : GuardResult()
}

/**
 * Route metadata
 */
data class RouteMeta(
    val title: String? = null,
    val requiresAuth: Boolean = false,
    val permissions: List<String> = emptyList()
)

/**
 * A single route definition
 */
data class Route(
    val path: String,
    val meta: RouteMeta = RouteMeta(),
    val guards: List<(RouteContext) -> GuardResult> = emptyList(),
    val children: List<Route> = emptyList()
)

/**
 * Context passed to route handlers
 */
class RouteContext(
    val url: ParsedURL,
    val router: ZylixRouter? = null
) {
    var isAuthenticated: Boolean = false
    var userRoles: List<String> = emptyList()
    var userData: Any? = null

    fun hasRole(role: String): Boolean = role in userRoles
}

// ============================================================================
// Navigation Event
// ============================================================================

enum class NavigationEvent {
    PUSH,
    REPLACE,
    BACK,
    FORWARD,
    DEEP_LINK
}

// ============================================================================
// Router
// ============================================================================

/**
 * Main router class for Zylix Android
 */
class ZylixRouter {
    private var routes: List<Route> = emptyList()
    private val history = mutableListOf<String>()
    private var historyIndex = -1
    private val navigationCallbacks = mutableListOf<(NavigationEvent, String, RouteContext) -> Unit>()
    private var notFoundHandler: ((ParsedURL) -> Unit)? = null
    private var basePath: String = ""

    private val _currentPath = MutableStateFlow("/")
    val currentPath: StateFlow<String> = _currentPath.asStateFlow()

    private val _currentContext = MutableStateFlow<RouteContext?>(null)
    val currentContext: StateFlow<RouteContext?> = _currentContext.asStateFlow()

    var navController: NavHostController? = null

    // ========================================================================
    // Configuration
    // ========================================================================

    fun defineRoutes(routes: List<Route>): ZylixRouter {
        this.routes = routes
        return this
    }

    fun setBasePath(path: String): ZylixRouter {
        this.basePath = path
        return this
    }

    fun onNotFound(handler: (ParsedURL) -> Unit): ZylixRouter {
        this.notFoundHandler = handler
        return this
    }

    fun onNavigate(callback: (NavigationEvent, String, RouteContext) -> Unit): ZylixRouter {
        navigationCallbacks.add(callback)
        return this
    }

    // ========================================================================
    // Navigation
    // ========================================================================

    fun push(path: String, userData: Any? = null) {
        navigate(path, NavigationEvent.PUSH, userData = userData)
    }

    fun replace(path: String, userData: Any? = null) {
        navigate(path, NavigationEvent.REPLACE, userData = userData)
    }

    fun back() {
        if (!canGoBack()) return
        historyIndex--
        val path = history[historyIndex]
        navigate(path, NavigationEvent.BACK, updateHistory = false)
    }

    fun forward() {
        if (!canGoForward()) return
        historyIndex++
        val path = history[historyIndex]
        navigate(path, NavigationEvent.FORWARD, updateHistory = false)
    }

    fun canGoBack(): Boolean = historyIndex > 0

    fun canGoForward(): Boolean = historyIndex < history.size - 1

    // ========================================================================
    // Deep Linking
    // ========================================================================

    fun handleDeepLink(uri: Uri) {
        val path = uri.path ?: "/"
        navigate(path, NavigationEvent.DEEP_LINK)
    }

    fun handleDeepLink(url: String) {
        val uri = Uri.parse(url)
        handleDeepLink(uri)
    }

    // ========================================================================
    // URL Parsing
    // ========================================================================

    fun parseURL(urlString: String): ParsedURL {
        var path = urlString
        var fragment: String? = null
        var queryString: String? = null

        // Extract fragment
        val hashIndex = path.indexOf('#')
        if (hashIndex >= 0) {
            fragment = path.substring(hashIndex + 1)
            path = path.substring(0, hashIndex)
        }

        // Extract query string
        val queryIndex = path.indexOf('?')
        if (queryIndex >= 0) {
            queryString = path.substring(queryIndex + 1)
            path = path.substring(0, queryIndex)
        }

        // Parse query parameters
        val query = mutableMapOf<String, String>()
        queryString?.split("&")?.forEach { pair ->
            val parts = pair.split("=", limit = 2)
            if (parts.size == 2) {
                query[parts[0]] = parts[1]
            }
        }

        return ParsedURL(path = path, query = query, fragment = fragment)
    }

    // ========================================================================
    // Route Matching
    // ========================================================================

    fun matchRoute(path: String): Pair<Route, List<RouteParam>>? {
        for (route in routes) {
            matchPattern(route.path, path)?.let { params ->
                return route to params
            }
            // Check children
            for (child in route.children) {
                val fullPattern = route.path + child.path
                matchPattern(fullPattern, path)?.let { params ->
                    return child to params
                }
            }
        }
        return null
    }

    private fun matchPattern(pattern: String, path: String): List<RouteParam>? {
        val patternParts = pattern.split("/").filter { it.isNotEmpty() }
        val pathParts = path.split("/").filter { it.isNotEmpty() }

        if (patternParts.size != pathParts.size) return null

        val params = mutableListOf<RouteParam>()

        for ((patternPart, pathPart) in patternParts.zip(pathParts)) {
            when {
                patternPart.startsWith(":") -> {
                    val paramName = patternPart.drop(1)
                    params.add(RouteParam(paramName, pathPart))
                }
                patternPart == "*" -> {
                    params.add(RouteParam("wildcard", pathPart))
                }
                patternPart != pathPart -> return null
            }
        }

        return params
    }

    // ========================================================================
    // Private Navigation
    // ========================================================================

    private fun navigate(
        path: String,
        event: NavigationEvent,
        updateHistory: Boolean = true,
        userData: Any? = null
    ) {
        val fullPath = basePath + path
        var parsed = parseURL(fullPath)

        // Match route
        val matched = matchRoute(parsed.path)
        if (matched == null) {
            notFoundHandler?.invoke(parsed)
            return
        }

        val (route, params) = matched

        // Update parsed URL with params
        parsed = parsed.copy(params = params)

        // Create context
        val context = RouteContext(parsed, this).apply {
            this.userData = userData
        }

        // Check guards
        for (guard in route.guards) {
            when (val result = guard(context)) {
                is GuardResult.Allow -> continue
                is GuardResult.Deny -> {
                    println("[ZylixRouter] Navigation denied: ${result.message}")
                    return
                }
                is GuardResult.Redirect -> {
                    replace(result.to)
                    return
                }
            }
        }

        // Update history
        if (updateHistory && (event == NavigationEvent.PUSH || event == NavigationEvent.DEEP_LINK)) {
            // Remove forward history
            if (historyIndex < history.size - 1) {
                history.subList(historyIndex + 1, history.size).clear()
            }
            history.add(path)
            historyIndex = history.size - 1
        }

        // Update state
        _currentPath.value = path
        _currentContext.value = context

        // Update NavController if available
        navController?.let { nav ->
            try {
                if (event == NavigationEvent.REPLACE) {
                    nav.popBackStack()
                }
                nav.navigate(path)
            } catch (e: Exception) {
                println("[ZylixRouter] NavController error: ${e.message}")
            }
        }

        // Notify callbacks
        for (callback in navigationCallbacks) {
            callback(event, path, context)
        }
    }
}

// ============================================================================
// Common Guards
// ============================================================================

fun requireAuth(context: RouteContext): GuardResult {
    return if (context.isAuthenticated) {
        GuardResult.Allow
    } else {
        GuardResult.Redirect("/login")
    }
}

fun requireRole(role: String): (RouteContext) -> GuardResult {
    return { context ->
        if (context.hasRole(role)) {
            GuardResult.Allow
        } else {
            GuardResult.Deny("Insufficient permissions")
        }
    }
}

// ============================================================================
// Compose Integration
// ============================================================================

/**
 * Remember and provide a ZylixRouter instance
 */
@Composable
fun rememberZylixRouter(): ZylixRouter {
    val navController = rememberNavController()
    return remember {
        ZylixRouter().also { it.navController = navController }
    }
}

/**
 * Composable to observe current route
 */
@Composable
fun ZylixRouter.observeCurrentPath(): State<String> {
    return currentPath.collectAsState()
}

/**
 * Composable to observe current context
 */
@Composable
fun ZylixRouter.observeCurrentContext(): State<RouteContext?> {
    return currentContext.collectAsState()
}

/**
 * Local composition for router access
 */
val LocalZylixRouter = staticCompositionLocalOf<ZylixRouter?> { null }

/**
 * Provider for router
 */
@Composable
fun ZylixRouterProvider(
    router: ZylixRouter,
    content: @Composable () -> Unit
) {
    CompositionLocalProvider(LocalZylixRouter provides router) {
        content()
    }
}

File: platforms/macos/Zylix/Sources/ZylixRouter.swift
// ZylixRouter.swift - macOS Router for Zylix v0.3.0
//
// Provides SwiftUI NavigationSplitView integration for Zylix routing system.
// Features:
// - NavigationSplitView integration for macOS
// - Deep link handling (Custom URL schemes)
// - Route parameters and query strings
// - Navigation guards
// - Window management integration

import SwiftUI
import Combine

// MARK: - Route Definition

/// A route parameter extracted from URL
public struct RouteParam: Equatable, Hashable {
    public let name: String
    public let value: String
}

/// Parsed URL components
public struct ParsedURL: Equatable {
    public let path: String
    public let params: [RouteParam]
    public let query: [String: String]
    public let fragment: String?

    public func getParam(_ name: String) -> String? {
        params.first { $0.name == name }?.value
    }

    public func getQuery(_ key: String) -> String? {
        query[key]
    }
}

/// Route guard result
public enum GuardResult {
    case allow
    case deny(message: String?)
    case redirect(to: String)
}

/// Route metadata
public struct RouteMeta {
    public var title: String?
    public var requiresAuth: Bool = false
    public var permissions: [String] = []
    public var showInSidebar: Bool = true
    public var icon: String?

    public init(
        title: String? = nil,
        requiresAuth: Bool = false,
        permissions: [String] = [],
        showInSidebar: Bool = true,
        icon: String? = nil
    ) {
        self.title = title
        self.requiresAuth = requiresAuth
        self.permissions = permissions
        self.showInSidebar = showInSidebar
        self.icon = icon
    }
}

/// A single route definition
public struct Route: Identifiable, Hashable {
    public let id = UUID()
    public let path: String
    public let meta: RouteMeta
    public var children: [Route] = []

    // Guards stored separately due to Hashable requirement
    internal var guardIds: [UUID] = []

    public init(
        path: String,
        meta: RouteMeta = RouteMeta(),
        children: [Route] = []
    ) {
        self.path = path
        self.meta = meta
        self.children = children
    }

    public static func == (lhs: Route, rhs: Route) -> Bool {
        lhs.id == rhs.id
    }

    public func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}

/// Context passed to route handlers
public class RouteContext: ObservableObject {
    @Published public var url: ParsedURL
    @Published public var isAuthenticated: Bool = false
    @Published public var userRoles: [String] = []
    public weak var router: ZylixRouter?
    public var userData: Any?

    public init(url: ParsedURL, router: ZylixRouter? = nil) {
        self.url = url
        self.router = router
    }

    public func hasRole(_ role: String) -> Bool {
        userRoles.contains(role)
    }
}

// MARK: - Navigation Event

public enum NavigationEvent {
    case push
    case replace
    case back
    case forward
    case deepLink
}

// MARK: - Router

@MainActor
public class ZylixRouter: ObservableObject {
    @Published public var currentPath: String = "/"
    @Published public var selectedRoute: Route?
    @Published public var currentContext: RouteContext?
    @Published public var sidebarRoutes: [Route] = []

    private var routes: [Route] = []
    private var guards: [UUID: (RouteContext) -> GuardResult] = [:]
    private var history: [String] = []
    private var historyIndex: Int = -1
    private var navigationCallbacks: [(NavigationEvent, String, RouteContext) -> Void] = []
    private var notFoundHandler: ((ParsedURL) -> Void)?
    private var basePath: String = ""

    public init() {}

    // MARK: - Configuration

    public func defineRoutes(_ routes: [Route]) -> ZylixRouter {
        self.routes = routes
        self.sidebarRoutes = routes.filter { $0.meta.showInSidebar }
        return self
    }

    public func addGuard(for route: inout Route, guard: @escaping (RouteContext) -> GuardResult) -> ZylixRouter {
        let guardId = UUID()
        route.guardIds.append(guardId)
        guards[guardId] = guard
        return self
    }

    public func setBasePath(_ path: String) -> ZylixRouter {
        self.basePath = path
        return self
    }

    public func onNotFound(_ handler: @escaping (ParsedURL) -> Void) -> ZylixRouter {
        self.notFoundHandler = handler
        return self
    }

    public func onNavigate(_ callback: @escaping (NavigationEvent, String, RouteContext) -> Void) -> ZylixRouter {
        navigationCallbacks.append(callback)
        return self
    }

    // MARK: - Navigation

    public func push(_ path: String, userData: Any? = nil) {
        navigate(path, event: .push, userData: userData)
    }

    public func replace(_ path: String, userData: Any? = nil) {
        navigate(path, event: .replace, userData: userData)
    }

    public func back() {
        guard canGoBack else { return }
        historyIndex -= 1
        let path = history[historyIndex]
        navigate(path, event: .back, updateHistory: false)
    }

    public func forward() {
        guard canGoForward else { return }
        historyIndex += 1
        let path = history[historyIndex]
        navigate(path, event: .forward, updateHistory: false)
    }

    public var canGoBack: Bool {
        historyIndex > 0
    }

    public var canGoForward: Bool {
        historyIndex < history.count - 1
    }

    public func select(_ route: Route) {
        push(route.path)
    }

    // MARK: - Deep Linking

    public func handleDeepLink(_ url: URL) {
        let path = url.path.isEmpty ? "/" : url.path
        navigate(path, event: .deepLink)
    }

    public func handleCustomURLScheme(_ url: URL) {
        // Handle zylix:// URLs
        guard url.scheme == "zylix" else { return }
        let path = "/" + url.host.map { $0 + url.path } .flatMap { $0 } ?? url.path
        navigate(path, event: .deepLink)
    }

    // MARK: - URL Parsing

    public func parseURL(_ urlString: String) -> ParsedURL {
        var path = urlString
        var fragment: String? = nil
        var queryString: String? = nil

        // Extract fragment
        if let hashIndex = path.firstIndex(of: "#") {
            fragment = String(path[path.index(after: hashIndex)...])
            path = String(path[..<hashIndex])
        }

        // Extract query string
        if let queryIndex = path.firstIndex(of: "?") {
            queryString = String(path[path.index(after: queryIndex)...])
            path = String(path[..<queryIndex])
        }

        // Parse query parameters
        var query: [String: String] = [:]
        if let qs = queryString {
            for pair in qs.split(separator: "&") {
                let parts = pair.split(separator: "=", maxSplits: 1)
                if parts.count == 2 {
                    query[String(parts[0])] = String(parts[1])
                }
            }
        }

        return ParsedURL(path: path, params: [], query: query, fragment: fragment)
    }

    // MARK: - Route Matching

    public func matchRoute(_ path: String) -> (Route, [RouteParam])? {
        for route in routes {
            if let params = matchPattern(route.path, path: path) {
                return (route, params)
            }
            for child in route.children {
                let fullPattern = route.path + child.path
                if let params = matchPattern(fullPattern, path: path) {
                    return (child, params)
                }
            }
        }
        return nil
    }

    private func matchPattern(_ pattern: String, path: String) -> [RouteParam]? {
        let patternParts = pattern.split(separator: "/", omittingEmptySubsequences: true)
        let pathParts = path.split(separator: "/", omittingEmptySubsequences: true)

        guard patternParts.count == pathParts.count else { return nil }

        var params: [RouteParam] = []

        for (patternPart, pathPart) in zip(patternParts, pathParts) {
            if patternPart.hasPrefix(":") {
                let paramName = String(patternPart.dropFirst())
                params.append(RouteParam(name: paramName, value: String(pathPart)))
            } else if patternPart == "*" {
                params.append(RouteParam(name: "wildcard", value: String(pathPart)))
            } else if patternPart != pathPart {
                return nil
            }
        }

        return params
    }

    // MARK: - Private Navigation

    private func navigate(_ path: String, event: NavigationEvent, updateHistory: Bool = true, userData: Any? = nil) {
        let fullPath = basePath + path
        var parsed = parseURL(fullPath)

        // Match route
        guard let (route, params) = matchRoute(parsed.path) else {
            notFoundHandler?(parsed)
            return
        }

        // Update parsed URL with params
        parsed = ParsedURL(path: parsed.path, params: params, query: parsed.query, fragment: parsed.fragment)

        // Create context
        let context = RouteContext(url: parsed, router: self)
        context.userData = userData

        // Check guards
        for guardId in route.guardIds {
            if let guard_ = guards[guardId] {
                let result = guard_(context)
                switch result {
                case .allow:
                    continue
                case .deny(let message):
                    print("[ZylixRouter] Navigation denied: \(message ?? "No reason")")
                    return
                case .redirect(let redirectPath):
                    replace(redirectPath)
                    return
                }
            }
        }

        // Update history
        if updateHistory && (event == .push || event == .deepLink) {
            if historyIndex < history.count - 1 {
                history.removeSubrange((historyIndex + 1)...)
            }
            history.append(path)
            historyIndex = history.count - 1
        }

        // Update state
        currentPath = path
        selectedRoute = route
        currentContext = context

        // Notify callbacks
        for callback in navigationCallbacks {
            callback(event, path, context)
        }
    }
}

// MARK: - Common Guards

public func requireAuth(_ context: RouteContext) -> GuardResult {
    if context.isAuthenticated {
        return .allow
    }
    return .redirect(to: "/login")
}

public func requireRole(_ role: String) -> (RouteContext) -> GuardResult {
    return { context in
        if context.hasRole(role) {
            return .allow
        }
        return .deny(message: "Insufficient permissions")
    }
}

// MARK: - SwiftUI Router View (macOS-specific with NavigationSplitView)

public struct MacOSRouterView<Sidebar: View, Detail: View>: View {
    @ObservedObject var router: ZylixRouter
    let sidebar: () -> Sidebar
    let detail: (RouteContext?) -> Detail

    public init(
        router: ZylixRouter,
        @ViewBuilder sidebar: @escaping () -> Sidebar,
        @ViewBuilder detail: @escaping (RouteContext?) -> Detail
    ) {
        self.router = router
        self.sidebar = sidebar
        self.detail = detail
    }

    public var body: some View {
        NavigationSplitView {
            sidebar()
        } detail: {
            detail(router.currentContext)
        }
        .onOpenURL { url in
            router.handleDeepLink(url)
        }
    }
}

// MARK: - Sidebar Navigation List

public struct RouterSidebarList: View {
    @ObservedObject var router: ZylixRouter

    public init(router: ZylixRouter) {
        self.router = router
    }

    public var body: some View {
        List(router.sidebarRoutes, selection: $router.selectedRoute) { route in
            NavigationLink(value: route) {
                Label(
                    route.meta.title ?? route.path,
                    systemImage: route.meta.icon ?? "folder"
                )
            }
        }
        .onChange(of: router.selectedRoute) { _, newValue in
            if let route = newValue {
                router.push(route.path)
            }
        }
    }
}

// MARK: - Router Environment

private struct RouterKey: EnvironmentKey {
    static let defaultValue: ZylixRouter? = nil
}

public extension EnvironmentValues {
    var zylixRouter: ZylixRouter? {
        get { self[RouterKey.self] }
        set { self[RouterKey.self] = newValue }
    }
}

public extension View {
    func zylixRouter(_ router: ZylixRouter) -> some View {
        environment(\.zylixRouter, router)
            .environmentObject(router)
    }
}
